cvmfs/quota.h

```diff
@@ -5,56 +5,437 @@
+  enum CommandType {
+    kTouch = 0,
+    kInsert,
+    kReserve,
+    kPin,
+    kUnpin,
+    kRemove,
+    kCleanup,
+    kList,
+    kListPinned,
+    kListCatalogs,
+    kStatus,
+    kLimits,
+    kPid,
+    kPinRegular,
+    kRegisterBackChannel,
+    kUnregisterBackChannel,
+    kGetProtocolRevision,
+    kInsertVolatile,
+    kListVolatile,
+  };
+
+  /**
+   * That could be done in more elegant way.  However, we might have a situation
+   * with old cache manager serving new clients (or vice versa) and we don't
+   * want to change the memory layout of LruCommand.
+   */
+  struct LruCommand {
+    CommandType command_type;
+    uint64_t size;    /**< Careful! Last 3 bits store hash algorithm */
+    int return_pipe;  /**< For cleanup, listing, and reservations */
+    unsigned char digest[shash::kMaxDigestSize];
+    /**
+     * Maximum 512-sizeof(LruCommand) in order to guarantee atomic pipe
+     * operations.
+     */
+    uint16_t desc_length;
+
+    LruCommand()
+      : command_type(static_cast<CommandType>(0))
+      , size(0)
+      , return_pipe(-1)
+      , desc_length(0)
+    {
+      memset(digest, 0, shash::kMaxDigestSize);
+    }
+
+    void SetSize(const uint64_t new_size) {
+      uint64_t mask = 7;
+      mask = ~(mask << (64-3));
+      size = (new_size & mask) | size;
+    }
+
+    uint64_t GetSize() const {
+      uint64_t mask = 7;
+      mask = ~(mask << (64-3));
+      return size & mask;
+    }
+
+    void StoreHash(const shash::Any &hash) {
+      memcpy(digest, hash.digest, hash.GetDigestSize());
+      // Exclude MD5
+      uint64_t algo_flags = hash.algorithm - 1;
+      algo_flags = algo_flags << (64-3);
+      size |= algo_flags;
+    }
+
+    shash::Any RetrieveHash() const {
+      uint64_t algo_flags = size >> (64-3);
+      shash::Any result(static_cast<shash::Algorithms>(algo_flags+1));
+      memcpy(result.digest, digest, result.GetDigestSize());
+      return result;
+    }
+  };
+
+  /**
+   * Maximum page cache per thread (Bytes).
+   */
+  static const unsigned kSqliteMemPerThread = 2*1024*1024;
+
+  /**
+   * Collect a number of insert and touch operations before processing them
+   * as sqlite commands.
+   */
+  static const unsigned kCommandBufferSize = 32;
+
+  /**
+   * Make sure that the amount of data transferred through the RPC pipe is
+   * within the OS's guarantees for atomiticity.
+   */
+  static const unsigned kMaxDescription = 512-sizeof(LruCommand);
+
+  /**
+   * Alarm when more than 75% of the cache fraction allowed for pinned files
+   * (50%) is filled with pinned files
+   */
+  static const unsigned kHighPinWatermark = 75;
+
+  /**
+   * The last bit in the sequence number indicates if an entry is volatile.
+   * Such sequence numbers are negative and they are preferred during cleanup.
+   * Volatile entries are used for instance for ALICE conditions data.
+   */
+  static const uint64_t kVolatileFlag = 1ULL << 63;
+
+
+  bool InitDatabase(const bool rebuild_database);
```

Comment:
Might be useful to encapsulate this into `sql::Database` since the quota database is the only one not managed through this abstract base class, to my knowledge.